pkg/noderesourcetopology/filter.go

```diff
@@ -152,19 +152,23 @@ func resourcesAvailableInAnyNUMANodes(logKey string, numaNodes NUMANodeList, res
 func isNUMANodeSuitable(qos v1.PodQOSClass, resource v1.ResourceName, quantity, numaQuantity resource.Quantity) bool {
 	// Check for the following:
 	if qos != v1.PodQOSGuaranteed {
-		// 1. set numa node as possible node if resource is memory or Hugepages
+		// 1. set numa node as possible node if resource is a non-native resources but sufficiently available on NUMA node(e.g. device)
```

Comment:
I got a bit confused by the code here. CMIIW, the code before this patch says for guaranteed pod, check for every resource type if current numa node is sufficient; for non guaranteed pod, ignore `memory`, `hugepage` and `cpu`, and check for the rest of the resource types if current numa node is sufficient. so for non guaranteed pod, the non-native resources are already checked?

Reply:
> for non guaranteed pod, ignore memory, hugepage and cpu, and check for the rest of the resource types if current numa node is sufficient. so for non guaranteed pod, the non-native resources are already checked?

(A) With non guaranteed pods the **previous** code devides into 2:
1. `BestEffort` pods - return immediately, IOW mark this node as a suitable one.
2.  `Burstable` pods -  `memory`, `hugepage` and `cpu` (AKA native resources) are not taken into account for `Burstable` pod so we don't need to check them at the NUMA level. And yes, native  resources already checked node-wide in the NodeResourceFit plugin which part of the native scheduler and prior to this plugin: https://github.com/kubernetes/kubernetes/blob/v1.25.2/pkg/scheduler/framework/plugins/noderesources/fit.go#L218 

(B) With non guaranteed pods **this** code has the same logic for both `BestEffort` and `Burstable` pods:
It checks whether a non-native device can be accounted for the pods, since unlike native resources, devices cannot 
  be shared among pods, even if they're not guaranteed ones.
  and then continues with step (A)2

I hope it addresses your question.

Reply:
Thanks for the explanation! I guess i'm on the same page now.
But according to the previous code, the non-native resources will be checked as well for `BestEffort` and `Burstable` pods I guess🤔? e.g. for `BestEffort` pods, the flow will fall through to L172  which is `return numaQuantity.Cmp(quantity) >= 0`, which means if I request devices in best effort pod, the quantity will be checked at numa level

Reply:
Not exactly :)
On the previous code,  for `BestEffort` pods we would immediately return here:
https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/pkg/noderesourcetopology/filter.go#L194

Reply:
aah sorry I missed that part xD. so `BestEffort` pod is not a proper example. But then it's the same thing with Burstable pod, i.e. for Burstable pod, if I request devices, the quantity is already checked by previous code? 

So to handle best effort and burstable pods with devices, I think we only have to remove https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/pkg/noderesourcetopology/filter.go#L194-L196? Actually if I comment out the changes in function `isNUMANodeSuitable`, the test case will still pass.

Reply:
hmm, you got a point.
I tend to agree, because if the resource is not one of the native resources it would fall off to the check on:
https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/pkg/noderesourcetopology/filter.go#L168

Reply:
Yes exactly!

Reply:
Thanks @pendoragon and @Tal-or, what you are saying is totally correct and is a great catch! 

As per @Huang-Wei's suggestion [here](https://github.com/kubernetes-sigs/scheduler-plugins/pull/425#discussion_r985297798), I will add a check and return early in case we have `BestEffort` pods without devices which would rule out a bunch of unwanted cases. Checks for requested quantity for both best effort pods and burstable pods would be handled properly automatically [here](https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/pkg/noderesourcetopology/filter.go#L168).

I will add this change as a separate commit for clarity and squash into the 1st commit once everyone is happy.

Thanks again for your input on this.